### ISSUE TITLE
single GPU version

### ISSUE THREAD

**Issue by melights** [mrharicot/monodepth#116]
Hi mrharicot,

Your code is great and I can train it on my machine with Titan X and 32GB RAM. I am trying to modify your training code to a single GPU version by removing the muti-gpu gradient average steps. Please see below for my modified code, however, it is automatically killed because of memory exhaustion after 100 steps (maybe in summary step?). But your original code works perfectly fine on my machine. I put anything else except the model in GPU like your orginal code. Do you have any idea why this happen? Many thanks!

`

    with tf.Graph().as_default(), tf.device('/cpu:0'):
        global_step = tf.Variable(0, trainable=False)
        # LEARNING RATE
        num_training_samples = count_text_lines(args.filenames_file)
        steps_per_epoch = np.ceil(num_training_samples / params.batch_size).astype(np.int32)
        num_total_steps = params.num_epochs * steps_per_epoch
        start_learning_rate = args.learning_rate
        boundaries = [np.int32((3/5) * num_total_steps), np.int32((4/5) * num_total_steps)]
        values = [args.learning_rate, args.learning_rate / 2, args.learning_rate / 4]
        learning_rate = tf.train.piecewise_constant(global_step, boundaries, values)
        
        #OPTIMIZER
        opt_step = tf.train.AdamOptimizer(learning_rate)
        print("total number of samples: {}".format(num_training_samples))
        print("total number of steps: {}".format(num_total_steps))

        # DATA
        dataloader = MonodepthDataloader(args.data_path, args.filenames_file, params, args.dataset, args.mode)
        left  = dataloader.left_image_batch
        right = dataloader.right_image_batch

        # MODEL
        with tf.device('/gpu:0'):
            model = MonodepthModel(params, args.mode, left, right, None)
            total_loss = model.total_loss
            train_op = opt_step.minimize(total_loss)

        # SUMMARY
        tf.summary.scalar('learning_rate', learning_rate, ['model_0'])
        tf.summary.scalar('total_loss', total_loss, ['model_0'])
        summary_op = tf.summary.merge_all('model_0')

        # SESSION
        config = tf.ConfigProto(allow_soft_placement=True)
        sess = tf.Session(config=config)

        # SAVER
        summary_writer = tf.summary.FileWriter(args.log_directory + '/' + args.model_name, sess.graph)
        train_saver = tf.train.Saver()

        # COUNT PARAMS 
        total_num_parameters = 0
        for variable in tf.trainable_variables():
            total_num_parameters += np.array(variable.get_shape().as_list()).prod()
        print("number of trainable parameters: {}".format(total_num_parameters))

        # INIT
        sess.run(tf.global_variables_initializer())
        sess.run(tf.local_variables_initializer())
        coordinator = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(sess=sess, coord=coordinator)
        #sess = tf_debug.LocalCLIDebugWrapperSession(sess)

        # LOAD CHECKPOINT IF SET
        if args.checkpoint_path != '':
            train_saver.restore(sess, args.checkpoint_path)
            
            if args.retrain:
                sess.run(global_step.assign(0))

        #TRAINING
        start_step = global_step.eval(session=sess)
        start_time = time.time()
        for step in range(start_step, num_total_steps):
            before_op_time = time.time()
            _, loss_value = sess.run([train_op, total_loss])
            duration = time.time() - before_op_time
            if step and step % 100 == 0:
                examples_per_sec = params.batch_size / duration
                time_sofar = (time.time() - start_time) / 3600
                training_time_left = (num_total_steps / step - 1.0) * time_sofar
                print_string = 'batch {:>6} | examples/s: {:4.2f} | loss: {:.5f} | time elapsed: {:.2f}h | time left: {:.2f}h'
                print(print_string.format(step, examples_per_sec, loss_value, time_sofar, training_time_left))
                summary_str = sess.run(summary_op)
                summary_writer.add_summary(summary_str, global_step=step)
`

**Comment by mrharicot** [comment]
Hi,
I don't see anything obvious on a first read.
Could you try putting the `train_op = opt_step.minimize(total_loss)` outside of the `with tf.device('/gpu:0'):` group?
